### PR TITLE
[recurbate.com] Add recurbate extractor

### PR DESCRIPTION
**Pull request by mrscrapy**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

This is a new extractor to support the NSFW website Recurbate.com, by default the website support 1 free download, so a premium subscription is required

Fixes #

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/recurbate.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            'title': 'Performer zsnicole33 show on 2022-10-25 20_23, Chaturbate Archive – Recurbate'
+        },
+        'skip': 'Free videos are available for a limited amount of time and for a single session.',
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._html_search_regex(r'<title>(.+?)</title>', webpage, 'title')
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:29]: _html_extract_title

File: yt_dlp/extractor/recurbate.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        token = self._html_search_regex(r'data-token=(.+?")', webpage, 'play_button').strip('"')
+        get_url = update_url_query('https://recurbate.com/api/get.php', {'video': video_id, 'token': token})
+        video_webpage = self._download_webpage(get_url, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:32]: Use `_download_webpage`'s query param

File: yt_dlp/extractor/recurbate.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        if 'shall_signin' in video_webpage[:20]:
+            raise ExtractorError(
+                "Login required: use --cookies to pass your browser's login cookie, or try again later",
+                expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:36]: raise_login_required

File: yt_dlp/extractor/recurbate.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    ExtractorError,
+    merge_dicts,
+    update_url_query,
+)
+
+from .common import InfoExtractor
+
+
+class RecurbateIE(InfoExtractor):
+    _VALID_URL = r'https?:\/\/(?:www\.)?recurbate\.com\/play\.php\?video=(?P<id>[0-9]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:14]: remove unnecessary escapes

File: yt_dlp/extractor/recurbate.py (hunk 3 above)
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:36]: do this check only if no entries are found

File: yt_dlp/extractor/recurbate.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        },
+        'skip': 'Free videos are available for a limited amount of time and for a single session.',
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._html_search_regex(r'<title>(.+?)</title>', webpage, 'title')
+        token = self._html_search_regex(r'data-token=(.+?")', webpage, 'play_button').strip('"')
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:30]:
```suggestion
        token = self._html_search_regex(r'data-token="([^"]+)"', webpage, 'token')
```

File: yt_dlp/extractor/recurbate.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            'id': '39161415',
+            'ext': 'mp4',
+            'title': 'Performer zsnicole33 show on 2022-10-25 20_23, Chaturbate Archive – Recurbate'
+        },
+        'skip': 'Free videos are available for a limited amount of time and for a single session.',
+    }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:27]: Is this necessary? Cant title be obtained from the other page?

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/recurbate.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+class RecurbateIE(InfoExtractor):
+    _VALID_URL = r'https?:\/\/(?:www\.)?recurbate\.com\/play\.php\?video=(?P<id>[0-9]+)'
+    _TEST = {
+        'url': 'https://recurbate.com/play.php?video=39161415',
+        'info_dict': {
+            'id': '39161415',
+            'ext': 'mp4',
+            'title': 'Performer zsnicole33 show on 2022-10-25 20_23, Chaturbate Archive – Recurbate'
+        },
+        'skip': 'Free videos are available for a limited amount of time and for a single session.',
```
Comment (pukkandan) [inline yt_dlp/extractor/recurbate.py:22]: Since each yt-dlp invocation is a new session, this shouldnt be an issue, correct?

**Comment by pukkandan** [comment]
* Many comments are still unaddressed
* https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#trailing-parentheses, https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#quotes
* Use of raise_login_required is wrong

File: yt_dlp/extractor/recurbate.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+from ..utils import ExtractorError, merge_dicts
+
+from .common import InfoExtractor
+
+
+class RecurbateIE(InfoExtractor):
+    _VALID_URL = r"https?://(?:www.)?recurbate.com/play.php\?video=(?P<id>[0-9]+)"
```
Comment (Lesmiscore) [inline yt_dlp/extractor/recurbate.py:7]: escape dots

File: yt_dlp/extractor/recurbate.py (hunk 7 above)
Comment (mrscrapy) [reply yt_dlp/extractor/recurbate.py:21]: I don't know how to explain this but it seems that the free download needs some time to become available again, and by not knowing how and when the tests are executed in this project, I cannot know if the test will be reliable or flaky at best.

Comment (pukkandan) [reply yt_dlp/extractor/recurbate.py:21]: > needs some time to become available again

Is it IP bound?

Comment (pukkandan) [reply yt_dlp/extractor/recurbate.py:21]: Multiple runs of the test are passing just fine
```
❯ pytest --tb=short -k Recurb
========================================================================================== test session starts ==========================================================================================
platform win32 -- Python 3.11.0, pytest-7.0.1, pluggy-1.0.0 -- D:\Programs\scoop\apps\python\current\python.exe
cachedir: .pytest_cache
rootdir: D:\Programs\Source\yt-dlp\yt-dlp, configfile: setup.cfg
collected 5249 items / 5247 deselected / 2 selected

test/test_download.py::TestDownload::test_Recurbate PASSED                                                                                                                                         [ 50%]
test/test_download.py::TestDownload::test_Recurbate_all PASSED                                                                                                                                     [100%]

================================================================================== 2 passed, 5247 deselected in 8.96s ===================================================================================

❯ pytest --tb=short -k Recurb
========================================================================================== test session starts ==========================================================================================
platform win32 -- Python 3.11.0, pytest-7.0.1, pluggy-1.0.0 -- D:\Programs\scoop\apps\python\current\python.exe
cachedir: .pytest_cache
rootdir: D:\Programs\Source\yt-dlp\yt-dlp, configfile: setup.cfg
collected 5249 items / 5247 deselected / 2 selected

test/test_download.py::TestDownload::test_Recurbate PASSED                                                                                                                                         [ 50%]
test/test_download.py::TestDownload::test_Recurbate_all PASSED                                                                                                                                     [100%]

================================================================================== 2 passed, 5247 deselected in 4.95s ===================================================================================

```

Comment (mrscrapy) [reply yt_dlp/extractor/recurbate.py]: no I believe there's an issue with the test, its downloading and validating a `403 forbidden` page

Comment (mrscrapy) [reply yt_dlp/extractor/recurbate.py]: Apologize, I see that this is how these tests are supposed to work, I expected those to download the full videos when running.

Are you sure you can reliably run the test without getting missing cookie exceptions?

Comment (pukkandan) [reply yt_dlp/extractor/recurbate.py:17]: I am not passing any `cookiefile`. If it breaks when using cookies, it should be considered a bug in the extractor

**Comment by pukkandan** [comment]
https://github.com/yt-dlp/yt-dlp/pull/5461/commits/e79adb1f4c1a42df997f4ed21637d0ea11b82954 what?

File: yt_dlp/extractor/recurbate.py (hunk 7 above)
Comment (mrscrapy) [reply yt_dlp/extractor/recurbate.py:17]: It is not failing when passing cookies, my question was related to how the tests are supposed to work.

My test fails after 3/4 runs

**Pull request closed without merging**